backend/shork/src/main/kotlin/internal/compiler/Scanner.kt

```diff
@@ -0,0 +1,173 @@
+        keywordMap["DIV"] = TokenType.DIV
+        keywordMap["MOD"] = TokenType.MOD
+        keywordMap["JMP"] = TokenType.JMP
+        keywordMap["JMZ"] = TokenType.JMZ
+        keywordMap["JMN"] = TokenType.JMN
+        keywordMap["DJN"] = TokenType.DJN
+        keywordMap["CMP"] = TokenType.CMP
+        keywordMap["SLT"] = TokenType.SLT
+        keywordMap["SPL"] = TokenType.SPL
+        keywordMap["ORG"] = TokenType.ORG
+        keywordMap["EQU"] = TokenType.EQU
+        keywordMap["END"] = TokenType.END
+        keywordMap["SEQ"] = TokenType.SEQ
+        keywordMap["SNE"] = TokenType.SNE
+        keywordMap["LDP"] = TokenType.LDP
+        keywordMap["STP"] = TokenType.STP
+        // Modifiers
+        keywordMap["A"] = TokenType.A
+        keywordMap["B"] = TokenType.B
+        keywordMap["AB"] = TokenType.AB
+        keywordMap["BA"] = TokenType.BA
+        keywordMap["F"] = TokenType.F
+        keywordMap["X"] = TokenType.X
+        keywordMap["I"] = TokenType.I
+    }
+
+    fun scanTokens(): List<Token> {
+        while (!isAtEnd()) {
+            // We are at the beginning of the next lexeme.
+            start = current
+            scanToken()
+        }
+
+        tokens.add(Token(TokenType.EOF, "", "", line))
+        return tokens
+    }
+
+    private fun scanToken() {
+        when (val c: Char = advance()) {
+            '+' -> addToken(TokenType.PLUS)
+            '-' -> addToken(TokenType.MINUS)
+            '*' -> addToken(TokenType.STAR)
+            '/' -> addToken(TokenType.SLASH)
+            ',' -> addToken(TokenType.COMMA)
+            '.' -> addToken(TokenType.DOT)
+            ';' -> {
+                addToken(TokenType.SEMICOLON)
+                while (peek() != '\n' && !isAtEnd()) {
+                    advance()
+                }
+            }
+            '{' -> addToken(TokenType.LEFT_BRACE)
+            '}' -> addToken(TokenType.RIGHT_BRACE)
+            '<' -> addToken(TokenType.LOWER_THAN)
+            '>' -> addToken(TokenType.GREATER_THAN)
+            '#' -> addToken(TokenType.HASHTAG)
+            '$' -> addToken(TokenType.DOLLAR)
+            '@' -> addToken(TokenType.AT)
+            ' ',
+            '\r',
+            '\t' -> {}
+            '\n' -> {
+                line++
+            }
+            else -> {
+                if (c.isDigit()) {
+                    number()
+                } else if (c.isLetter()) {
+                    identifier()
+                } else {
+                    errors.add(Pair(line, "Unexpected character: ${c} at pos $current"))
+                }
+            }
+        }
+    }
+
+    private fun advance(): Char {
+        return source[current++]
+    }
+
+    private fun peek(): Char {
+        if (isAtEnd()) {
+            return '\u0000'
+        }
+        return source[current]
+    }
+
+    /**
+     * Used for lexemes like `++`, `--`, `==`, `!=`, `<=`, `>=`.
+     *
+     * Basically a conditional advance()
+     */
```

Comment:
Make this comment a bit clearer, describe what the match function actually does.